### PR TITLE
[WIP] Improve RichPayloadEventSource

### PR DESCRIPTION
**Pull request by MS-TimothyMothra**
proposed changes for #998 

DO NOT MERGE
This PR causes problems that our unit tests currently do not catch

### CODE REVIEW

File: src/Microsoft.ApplicationInsights/DataContracts/ISupportRichPayloadEventSource.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+﻿namespace Microsoft.ApplicationInsights.DataContracts
+{
+    using System.Diagnostics.Tracing;
+
+    /// <summary>
+    ///
```
Comment (pharring) [inline src/Microsoft.ApplicationInsights/DataContracts/ISupportRichPayloadEventSource.cs]: Fill in the summary text. Something like: "Marker interface to indicate that a telemetry item supports serializing to an event source."

**Comment by pharring** [comment]
The change looks good to me, but we need to be pretty careful about changes to RichPayloadEventSource. The last round of changes caused a regression in AI Profiler due to a change in serialization of "Duration". In this PR, there's a subtle change in the WriteEvent method -- 'data' used to be strongly-typed, now it's just an object. I don't _think_ it'll be a problem, but I'm not 100% certain. 
I want to add some new unit tests in this repo to try and catch serialization changes -- I intended to do this already, but I've been busy with other things.

File: src/Microsoft.ApplicationInsights/Extensibility/Implementation/RichPayloadEventSource.cs [hunk 2]
```diff
@@ -40,209 +40,25 @@ public RichPayloadEventSource()
...
-                CopyGlobalPropertiesIfRequired(item, telemetryItem.Properties);
-                item.Sanitize();
-                this.WriteEvent(
-                    AvailabilityTelemetry.TelemetryName,
-                    telemetryItem.Context.InstrumentationKey,
-                    telemetryItem.Context.SanitizedTags,
-                    telemetryItem.Data,
-                    telemetryItem.Context.Flags,
-                    Keywords.Availability);
+                    eventName: RequestTelemetry.TelemetryName,
```
Comment (pharring) [inline src/Microsoft.ApplicationInsights/Extensibility/Implementation/RichPayloadEventSource.cs]: This should be `eventName: richPayloadTelemetryItem.TelemetryName`

**Comment by cijothomas** [comment]
@pharring Please do add unit tests when you get a chance to ensure we dont break AI Profiler!

**Comment by MS-TimothyMothra** [comment]
@pharring @cijothomas i can look into unit tests. I would like the confirmation that my changes don't break anything.

File: src/Microsoft.ApplicationInsights/Extensibility/Implementation/RichPayloadEventSource.cs (hunk 2 above)
Comment (MS-TimothyMothra) [reply src/Microsoft.ApplicationInsights/Extensibility/Implementation/RichPayloadEventSource.cs]: thank you

File: src/Microsoft.ApplicationInsights/DataContracts/DependencyTelemetry.cs [hunk 3]
```diff
@@ -104,6 +105,12 @@ private DependencyTelemetry(DependencyTelemetry source)
         /// <inheritdoc />
         string IAiSerializableTelemetry.BaseType => nameof(RemoteDependencyData);
 
+        EventKeywords ISupportRichPayloadEventSource.EventSourceKeyword => RichPayloadEventSource.Keywords.Dependencies;
+
+        object ISupportRichPayloadEventSource.Data => this.Data;
```
Comment (cijothomas) [inline src/Microsoft.ApplicationInsights/DataContracts/DependencyTelemetry.cs]: This is not correct.. It should be this.InternalData

Comment (cijothomas) [reply src/Microsoft.ApplicationInsights/DataContracts/DependencyTelemetry.cs]: (in this case .Data is actually the field to store sqlcommand etc) name is confusing.

File: src/Microsoft.ApplicationInsights/DataContracts/ExceptionTelemetry.cs [hunk 4]
```diff
@@ -100,6 +101,12 @@ private ExceptionTelemetry(ExceptionTelemetry source)
         /// <inheritdoc />
         string IAiSerializableTelemetry.BaseType => nameof(ExceptionData);
 
+        EventKeywords ISupportRichPayloadEventSource.EventSourceKeyword => RichPayloadEventSource.Keywords.Exceptions;
+
+        object ISupportRichPayloadEventSource.Data => this.Data;
```
Comment (cijothomas) [inline src/Microsoft.ApplicationInsights/DataContracts/ExceptionTelemetry.cs]: this.Data.Data

**Comment by alohaninja** [comment]
Trying to revive this thread - we are seeing intense performance issues (_2-5s per log request_) with `RichPayloadEventSource` - how can we turn this feature off or escalate a fix? This is a blocking call and causing significant performance delays for us using the serilog sink for AI. 

image

**Comment by pharring** [comment]
@alohaninja are you sure you're interpreting that trace correctly? It looks like the time is spent in JIT compilation which should happen only once. The changes in this PR wouldn't, I think fix that.

For sure 2.5 seconds of JIT compilation time is high, but does it happen on _every_ request?

**Comment by alohaninja** [comment]
@pharring - this issue does seem to appear with each request - not familiar with the action that's occurring when `RichPayloadEventSource` fires. Is it required to use AI? it does seem to be impacting multiple AI modules. 

image

**Pull request closed without merging**